jina/executors/indexers/vector.py

```diff
@@ -114,16 +116,42 @@ def _validate_key_vector_shapes(self, keys, vectors):
     def add(self, keys: 'np.ndarray', vectors: 'np.ndarray', *args, **kwargs) -> None:
         self._validate_key_vector_shapes(keys, vectors)
         self.write_handler.write(vectors.tobytes())
+        self.valid_indices = np.concatenate((self.valid_indices, np.full(len(keys), True)))
         self.key_bytes += keys.tobytes()
         self.key_dtype = keys.dtype.name
         self._size += keys.shape[0]
 
+    def _check_keys(self, keys: Sequence[int]) -> None:
+        missed = []
+        for key in keys:
+            # if it never existed or if it's been marked as deleted in the current index
+            # using `is False` doesn't work
+            if key not in self.ext2int_id.keys() or self.valid_indices[self.ext2int_id[key]] == False:  # noqa
+                missed.append(key)
+        if missed:
+            raise KeyError(f'Key(s) {missed} were not found in {self.save_abspath}')
+
+    def update(self, keys: Sequence[int], values: Sequence[bytes], *args, **kwargs) -> None:
+        self.delete(keys)
+        self.add(np.array(keys), np.array(values))
+
+    def delete(self, keys: Sequence[int], *args, **kwargs) -> None:
+        self._check_keys(keys)
+        for key in keys:
+            # mark as `False` in mask
+            self.valid_indices[self.ext2int_id[key]] = False
+        self._size = self.size - len(list(keys))
+
     def get_query_handler(self) -> Optional['np.ndarray']:
         """Open a gzip file and load it as a numpy ndarray
 
         :return: a numpy ndarray of vectors
         """
-        vecs = self.raw_ndarray
+        if np.all(self.valid_indices is True):
+            vecs = self.raw_ndarray
```

Comment:
I think this line would be important to cover in the tests

Reply:
this is covered for sure

Reply:
When I  run the 47 tests of the test_numpyindexer.py locally, this line is not called.

Reply:
Sure, weird, will fix

Reply:
Can u check now?

Reply:
you can change it to `np.all(self.valid_indices)`

Reply:
done!